src/app/rpmostree-compose-builtin-tree.c

```diff
@@ -1021,10 +1047,157 @@ impl_compose_tree (const char      *treefile_pathstr,
+  g_print ("rootfs: %s/rootfs\n", self->workdir_tmp.path);
+  glnx_tmpdir_unset (&self->workdir_tmp);
 
-  return TRUE;
+  return EXIT_SUCCESS;
+}
+
+int
+rpmostree_compose_builtin_postprocess_root (int             argc,
+                                            char          **argv,
+                                            RpmOstreeCommandInvocation *invocation,
+                                            GCancellable   *cancellable,
+                                            GError        **error)
+{
+  g_autoptr(GOptionContext) context = g_option_context_new ("postprocess ROOTFS [TREEFILE]");
+  if (!rpmostree_option_context_parse (context,
+                                       postprocess_option_entries,
+                                       &argc, &argv,
+                                       invocation,
+                                       cancellable,
+                                       NULL, NULL, NULL, NULL,
+                                       error))
+    return EXIT_FAILURE;
+
+  if (argc < 2 || argc > 3)
+    {
+      rpmostree_usage_error (context, "ROOTFS must be specified", error);
+      return EXIT_FAILURE;
+    }
+
+  const char *rootfs_path = argv[1];
+  /* Here we *optionally* process a treefile; some things like `tmp-is-dir` and
+   * `boot_location` are configurable and relevant here, but a lot of users
+   * will also probably be OK with the defaults, and part of the idea here is
+   * to avoid at least some of the use cases requiring a treefile.
+   */
+  const char *treefile_path = argc > 2 ? argv[2] : NULL;
+  glnx_unref_object JsonParser *treefile_parser = NULL;
+  JsonObject *treefile = NULL; /* Owned by parser */
+  if (treefile_path)
+    {
+      treefile_parser = json_parser_new ();
+      if (!json_parser_load_from_file (treefile_parser, treefile_path, error))
+        return EXIT_FAILURE;
+
+      JsonNode *treefile_rootval = json_parser_get_root (treefile_parser);
+      if (!JSON_NODE_HOLDS_OBJECT (treefile_rootval))
+        return glnx_throw (error, "Treefile root is not an object"), EXIT_FAILURE;
+      treefile = json_node_get_object (treefile_rootval);
+    }
+
+  glnx_fd_close int rootfs_dfd = -1;
+  if (!glnx_opendirat (AT_FDCWD, rootfs_path, TRUE, &rootfs_dfd, error))
+    return EXIT_FAILURE;
+  if (!rpmostree_rootfs_postprocess_common (rootfs_dfd, cancellable, error))
+    return EXIT_FAILURE;
+  if (!rpmostree_postprocess_final (rootfs_dfd, treefile,
+                                    cancellable, error))
+    return EXIT_FAILURE;
+  return EXIT_SUCCESS;
+}
+
+int
+rpmostree_compose_builtin_commit_root (int             argc,
+                                       char          **argv,
+                                       RpmOstreeCommandInvocation *invocation,
+                                       GCancellable   *cancellable,
+                                       GError        **error)
+{
+  g_autoptr(GOptionContext) context = g_option_context_new ("TREEFILE");
```

Comment:
hmm, do we need a ROOTFS here in the description? Otherwise, if we don't have rootfs, we will fail at the line  https://github.com/projectatomic/rpm-ostree/pull/1039/files#diff-76a6c42abcbc88a41dc96851b2af87d2R1196 right?

Reply:
Yep, fixed!